File: digital_wallet_fastapi/app/services/auth_service.py
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import hash_password, verify_password, create_access_token
from app.exceptions.handlers import AuthenticationException, DuplicateUsernameException
from app.models.user import User
from app.models.wallet import Wallet
from app.schemas.auth import LoginResponse, UserResponse


async def register(session: AsyncSession, username: str, password: str) -> User:
    # role is always ROLE_USER, ignored from client input
    user = User(username=username, password_hash=hash_password(password), role="ROLE_USER")
    session.add(user)
    try:
        await session.flush()
    except IntegrityError:
        raise DuplicateUsernameException(f"Username '{username}' is already taken")

    # create wallet
    wallet = Wallet(user_id=user.id, currency="USDT", balance=0, version=0)
    session.add(wallet)

    return user


async def login(session: AsyncSession, username: str, password: str) -> LoginResponse:
    stmt = select(User).where(User.username == username)
    result = await session.execute(stmt)
    user = result.scalar_one_or_none()

    if user is None:
        raise AuthenticationException("Invalid username or password")

    if not verify_password(password, user.password_hash):
        raise AuthenticationException("Invalid username or password")

    token = create_access_token(user.id, user.username)

    user_response = UserResponse(
        id=user.id,
        username=user.username,
        role=user.role,
        createdAt=user.created_at,
    )

    return LoginResponse(token=token, user=user_response)

File: digital_wallet_laravel/app/Services/AuthService.php
<?php

namespace App\Services;

use App\Exceptions\AuthenticationException;
use App\Exceptions\DuplicateUsernameException;
use App\Helpers\JwtHelper;
use App\Models\User;
use App\Models\Wallet;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Hash;

class AuthService
{
    public function register(string $username, string $password): User
    {
        return DB::transaction(function () use ($username, $password) {
            try {
                $user = User::create([
                    'username' => $username,
                    'password_hash' => Hash::make($password),
                    'role' => 'ROLE_USER',
                ]);
            } catch (\Illuminate\Database\UniqueConstraintViolationException $e) {
                throw new DuplicateUsernameException("Username '{$username}' is already taken");
            }

            Wallet::create([
                'user_id' => $user->id,
                'currency' => 'USDT',
                'balance' => 0,
                'version' => 0,
            ]);

            return $user;
        });
    }

    public function login(string $username, string $password): array
    {
        $user = User::where('username', $username)->first();

        if (!$user) {
            throw new AuthenticationException('Invalid username or password');
        }

        if (!Hash::check($password, $user->password_hash)) {
            throw new AuthenticationException('Invalid username or password');
        }

        $token = JwtHelper::generateToken($user->id, $user->username);

        return [
            'token' => $token,
            'user' => [
                'id' => $user->id,
                'username' => $user->username,
                'role' => $user->role,
                'createdAt' => $user->created_at?->toISOString(),
            ],
        ];
    }
}
